### PR TITLE
ci(invariants): enforce invariant-to-test coverage mapping

### PR DESCRIPTION
**Pull request by hnipps**
## Summary

Closes item **#01** of the deterministic-assurance roadmap
([docs/assurance/immediate/01-invariant-test-coverage-ci.md](../blob/main/docs/assurance/immediate/01-invariant-test-coverage-ci.md)).

Makes the governance rule in `docs/invariants/*.md` (every declared invariant has a covering property test, and every test comment refers to a declared invariant) **mechanically enforced** in pre-commit and CI.

## What this adds

- **`scripts/check_invariant_coverage.py`** (195 LOC) — enforces bidirectional mapping:
  - Doc headers: `**I1. Name.**` (regex `^\*\*([A-Z]+\d+[a-z]?)\.\s`)
  - Test comments: `// Invariant I1: Name.` (regex `^//\s*Invariant\s+([A-Z]+\d+[a-z]?):\s`)
  - Scoped **per-module** (queue's `I1` ≠ runner's `I1`).
  - Aspirational opt-out via `<!-- aspirational -->` on the header line.
  - Exit codes: `0` clean, `1` violations, `2` malformed inputs.
- **Pre-commit hook** `invariant-coverage` (local, `language: system`).
- **CI job** `invariant-coverage` parallel to `go-checks`; widened workflow `paths` to trigger on `docs/invariants/**` and the script itself.
- **`docs/invariants/README.md`** — documents the enforcement mechanism, the aspirational opt-out, and the add/remove workflows.

## Empirical validation

Three induced failures on `cli/internal/queue/queue_invariants_prop_test.go` reproduced the three acceptance-criterion failure modes from the roadmap. All reverted before commit.

1. **Rename comment `I1 → I99`:**
   ```
   invariant coverage check FAILED:
     docs/invariants/queue.md: invariant I1 is declared but has no `// Invariant I1: ...` comment in cli/internal/queue/*invariants_prop_test.go
     docs/invariants/queue.md: `// Invariant I99:` refers to an invariant not declared in this module doc (sites: cli/internal/queue/queue_invariants_prop_test.go:214)
   ```

2. **Delete the `// Invariant I1:` comment:**
   ```
   invariant coverage check FAILED:
     docs/invariants/queue.md: invariant I1 is declared but has no `// Invariant I1: ...` comment in cli/internal/queue/*invariants_prop_test.go
   ```

3. **Add an orphan comment for undeclared ID `I99`:**
   ```
   invariant coverage check FAILED:
     docs/invariants/queue.md: `// Invariant I99:` refers to an invariant not declared in this module doc (sites: cli/internal/queue/queue_invariants_prop_test.go:215)
   ```

Clean baseline: `invariant coverage OK: 35/35 required invariants covered across 3 module(s) (0 aspirational).`

Runtime: 30–50 ms (budget: 10 s).

## Design notes

- **Protected-surfaces resolution.** The roadmap originally listed edits to `docs/invariants/queue.md § Governance`, which is on the protected-surfaces list. Deliverable §4 explicitly allowed a `README.md` alternative — that is what this PR takes. `queue.md § Governance` remains untouched; the README is the pointer.
- **Non-digit ID typo behaviour.** The regex (`[A-Z]+\d+[a-z]?`) requires a digit. A typo that drops the digit (e.g. `I1 → Ix`) is invisible to the regex: the check surfaces it as **missing coverage on the original ID**, not as an orphan on the malformed label. This is intentional (reject guesses) and is called out in `docs/invariants/README.md § Comment format`.
- **Scope.** Two new untracked files + two edits to existing tracked files. No Go source changes; no behavioural changes to xylem. Fails closed on violations.

## Test plan

- [x] Script passes on clean tree (35/35 across queue/runner/scanner)
- [x] Three induced-failure modes from roadmap acceptance criteria reproduced and diagnosed correctly
- [x] Both YAML files parse (`yaml.safe_load`)
- [x] Pre-commit hook wired and runs: `pre-commit run invariant-coverage --all-files` → `Passed`
- [x] Pre-commit runs the hook on this very commit (see commit hook output)
- [ ] CI job passes on this PR (watch `invariant-coverage` job)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by hnipps** [comment]
**xylem — phase `analyze` completed** (2m47.212349s)

_Usage:_ $0.0954, 4223

<details>
<summary>Phase output (click to expand)</summary>

gh API is fully rate-limited this session. I have everything I need from the local git data. Here's the analysis:

---

## PR #665 Review Analysis

**Summary**: Adds a Python script (`check_invariant_coverage.py`) that enforces a bidirectional mapping between invariant specs in `docs/invariants/*.md` and `// Invariant <ID>:` comments in property test files, wired into both pre-commit and CI.

**Changed files**:
| File | Type |
|---|---|
| `.github/workflows/ci.yml` | modified |
| `.pre-commit-config.yaml` | modified |
| `docs/invariants/README.md` | added |
| `scripts/check_invariant_coverage.py` | added |

**Test files changed**: None (no `*_test.go` files in diff).

**Intent**: Closes assurance roadmap item #01 — mechanically enforce the governance rule that every declared module invariant has a covering property test (and vice versa), preventing silent drift.

**Scope assessment**: Well-scoped. Four files, no Go code changes, no behavioral changes to xylem. Strictly additive enforcement tooling.

**Anchor comment ID**: Unable to determine (gh API rate-limited/403).

**Existing inline comments**: Unable to determine (gh API rate-limited/403).

---

### Key finding: CI path-filter invariant violation

The PR adds `docs/invariants/**` and `scripts/check_invariant_coverage.py` to `ci.yml`'s `paths:` list but does **not** add matching entries to `ci-skip.yml`'s `paths-ignore:` list. Per CLAUDE.md:

> **Invariant:** the `paths:` list in `ci.yml` and the `paths-ignore:` list in `ci-skip.yml` must be identical at all times.

Current `ci-skip.yml` `paths-ignore` (3 entries):
- `cli/**`
- `.golangci.yml`
- `.xylem/workflows/**`

PR's `ci.yml` `paths` (5 entries):
- `cli/**`
- `.golangci.yml`
- `.xylem/workflows/**`
- `docs/invariants/**` **(missing from ci-skip)**
- `scripts/check_invariant_coverage.py` **(missing from ci-skip)**

This means PRs touching only `docs/invariants/` or the script will trigger *neither* `ci.yml` *nor* `ci-skip.yml`, leaving the required `go-checks` status permanently pending and blocking merge. The `invariant-coverage` job would also not run since it shares the same `on:` trigger.


</details>

**Comment by hnipps** [comment]
**xylem — phase `analyze` completed** (4m20.499866s)

_Usage:_ $0.0941, 4205

<details>
<summary>Phase output (click to expand)</summary>

---

## PR #665 Review — Read-Only Analysis

### Summary

Adds a Python script (`check_invariant_coverage.py`) that enforces bidirectional mapping between invariant declarations in `docs/invariants/*.md` and `// Invariant <ID>:` comments in property test files, wired into both pre-commit and CI. Closes assurance roadmap item #01.

### Changed files

| File | Type |
|---|---|
| `.github/workflows/ci.yml` | Modified |
| `.pre-commit-config.yaml` | Modified |
| `docs/invariants/README.md` | Added |
| `scripts/check_invariant_coverage.py` | Added |

### Test files changed

None — no `*_test.go` files in the diff.

### Intent

Mechanically enforce the governance rule that every declared module invariant has a covering property test comment, and vice versa. Prevents silent drift between specs and tests.

### Scope assessment

Well-scoped: 4 files, 302 lines added, no Go source changes, no behavioral changes to xylem. Script is fail-closed on violations (exit 1) and malformed input (exit 2). File mode is correctly 100755. Aspirational opt-out regex is non-DOTALL (safe). `discover_modules()` correctly skips `README.md` case-insensitively. Missing test-file directory case produces a clear "no property test file found" error rather than silently passing.

**BLOCKER: `ci-skip.yml` paths-ignore not updated.** The PR adds `docs/invariants/**` and `scripts/check_invariant_coverage.py` to `ci.yml`'s `paths:` list but does **not** add them to `.github/workflows/ci-skip.yml`'s `paths-ignore:` list. Per the CLAUDE.md invariant ("the `paths:` list in `ci.yml` and the `paths-ignore:` list in `ci-skip.yml` must be identical at all times"), this is a violation. A docs/invariants-only change would trigger both the real `go-checks` job (ci.yml matches) and the no-op `go-checks` job (ci-skip.yml doesn't exclude it), producing duplicate status checks.

### Anchor comment ID

Unable to verify — GitHub API blocked under current sandbox restrictions.

### Existing inline comments

Unable to verify — GitHub API blocked under current sandbox restrictions.


</details>

**Comment by hnipps** [comment]
**xylem — phase `verify` completed** (5m24.543361s)

_Usage:_ $0.2105, 6146

<details>
<summary>Phase output (click to expand)</summary>

I have all the evidence needed. Here's the structured verification report.

---

## Verification Report — PR #665

```
QUESTION: Are the code changes in PR #665 correct?
SCOPE: .github/workflows/ci.yml, .github/workflows/ci-skip.yml,
       .pre-commit-config.yaml, scripts/check_invariant_coverage.py,
       docs/invariants/README.md
```

### Premises

```
PREMISE P1 [STATIC]: ci.yml paths list now contains 5 entries:
  cli/**, .golangci.yml, .xylem/workflows/**, docs/invariants/**,
  scripts/check_invariant_coverage.py
  Evidence: ci.yml:7-11 (push), ci.yml:14-18 (pull_request) — PR diff

PREMISE P2 [STATIC]: ci-skip.yml paths-ignore list still contains only 3 entries:
  cli/**, .golangci.yml, .xylem/workflows/**
  Evidence: ci-skip.yml:6-9 — git show pr-665-review:.github/workflows/ci-skip.yml

PREMISE P3 [STATIC]: CLAUDE.md documents a mandatory invariant: "the paths: list
  in ci.yml and the paths-ignore: list in ci-skip.yml must be identical at all times"
  Evidence: CLAUDE.md § "CI path-filter invariant"

PREMISE P4 [STATIC]: HEADER_RE correctly matches all existing invariant headers
  in queue.md (13 IDs), runner.md (14 IDs), scanner.md (8 IDs) while rejecting
  non-invariant bold lines (**Summary:**, **Amendment procedure:**, etc.)
  Evidence: Verified via Python regex test against all 18 bold lines in the 3 docs

PREMISE P5 [STATIC]: TEST_COMMENT_RE correctly matches all existing test comments
  (13 in queue, 14 in runner, 8 in scanner) while rejecting prose mentions
  and section headers ("// Invariant properties")
  Evidence: Verified via Python regex test against 7 representative lines

PREMISE P6 [STATIC]: Running the script's logic against current repo state produces
  zero violations — all 35 declared invariants have matching test comments, zero orphans
  Evidence: Simulated execution printing declared/covered sets for all 3 modules

PREMISE P7 [STATIC]: REPO_ROOT is computed as Path(__file__).resolve().parent.parent,
  which correctly resolves to the repo root since the script lives at scripts/
  Evidence: check_invariant_coverage.py:39

PREMISE P8 [STATIC]: The CI job invariant-coverage runs from repo root (no
  working-directory override), matching the script's REPO_ROOT computation
  Evidence: ci.yml:50-55 — no defaults.run.working-directory set

PREMISE P9 [STATIC]: discover_modules() uses glob("*.md") and skips README
  case-insensitively via stem.lower() == "readme"
  Evidence: check_invariant_coverage.py:77-79

PREMISE P10 [STATIC]: The pre-commit files pattern
  ^(docs/invariants/.+\.md|cli/internal/.+_invariants_prop_test\.go|scripts/check_invariant_coverage\.py)$
  correctly matches all relevant files and excludes unrelated ones
  Evidence: Verified via Python regex test against 9 representative paths

PREMISE P11 [STATIC]: The discover_modules glob *invariants_prop_test.go is slightly
  looser than the pre-commit regex _invariants_prop_test.go (would match a hypothetical
  file named just invariants_prop_test.go), but all 3 existing files follow the
  <module>_invariants_prop_test.go convention
  Evidence: find cli/internal -name '*invariants_prop_test.go' returns 3 files,
  all with module prefix
```

### Claims

```
CLAIM C1 [STATIC]: The ci.yml/ci-skip.yml path-filter invariant is violated.
  Trace: ci.yml adds 2 new paths (P1) → ci-skip.yml unchanged (P2)
         → lists no longer identical → violates CLAUDE.md invariant (P3)
  Depends on: P1, P2, P3

CLAIM C2 [STATIC]: A PR touching ONLY docs/invariants/** would trigger BOTH
  ci.yml (path matches) AND ci-skip.yml (path NOT in paths-ignore), producing
  duplicate go-checks status contexts.
  Trace: docs/invariants/queue.md edit → ci.yml paths match "docs/invariants/**" → fires
         → ci-skip.yml paths-ignore doesn't include "docs/invariants/**" → also fires
         → two go-checks jobs
  Depends on: P1, P2

CLAIM C3 [STATIC]: The coverage script is correct for all existing modules — its
  regex patterns, discovery logic, and scoping are sound.
  Trace: HEADER_RE (P4) × 3 docs → TEST_COMMENT_RE (P5) × 3 test files
         → 35/35 covered, 0 orphans (P6)
  Depends on: P4, P5, P6, P7, P9

CLAIM C4 [STATIC]: The CI job will execute correctly — working directory and
  path resolution are consistent.
  Depends on: P7, P8
```

### Alternative Hypothesis Check

```
ALTERNATIVE HYPOTHESIS CHECK:
If the ci-skip.yml gap were harmless, what evidence would exist?
- Searched for: any mechanism that prevents ci-skip.yml from firing when ci.yml fires
- Found: GitHub Actions path filters operate independently per workflow file;
  there is no cross-workflow suppression. Both can fire on the same PR.
- Conclusion: REFUTED — the gap produces real duplicate status checks

ALTERNATIVE HYPOTHESIS CHECK:
If the coverage script had a false-negative (passing when it shouldn't), what evidence?
- Searched for: invariant IDs declared in docs that lack test comments
- Found: All 35 declared IDs have matching test comments (P6)
- Searched for: test comments referencing undeclared IDs
- Found: None — all comment IDs map to declared invariants
- Conclusion: REFUTED — the script correctly passes on current state
```

---

## Findings

### Finding 1

- **File and line**: `.github/workflows/ci-skip.yml:6-9`
- **Severity**: CRITICAL
- **Category**: Invariant violation
- **Description**: The `paths-ignore:` list in `ci-skip.yml` is missing `docs/invariants/**` and `scripts/check_invariant_coverage.py`, which were added to `ci.yml`'s `paths:` list. This violates the documented invariant ("the `paths:` list in `ci.yml` and the `paths-ignore:` list in `ci-skip.yml` must be identical at all times"). A docs/invariants-only PR will fire both workflows, producing duplicate `go-checks` status contexts.
- **Evidence**: `ci.yml:7-11` adds 5 paths; `ci-skip.yml:6-9` has only 3 (the original set). CLAUDE.md explicitly documents this as a mandatory invariant.

### Finding 2

- **File and line**: `scripts/check_invariant_coverage.py:80`
- **Severity**: MEDIUM
- **Category**: Edge case
- **Description**: The glob `*invariants_prop_test.go` in `discover_modules()` would match a hypothetical file named `invariants_prop_test.go` (without module prefix), while the pre-commit `files` regex requires `_invariants_prop_test.go` (with underscore). This is a minor inconsistency — no such file exists today, and the script would still function correctly, but the glob could be tightened to `*_invariants_prop_test.go` for consistency.
- **Evidence**: `check_invariant_coverage.py:80` uses `module_dir.glob("*invariants_prop_test.go")`; `.pre-commit-config.yaml:68` uses `cli/internal/.+_invariants_prop_test\.go`.

---

```
FORMAL CONCLUSION:
By C1 (P1, P2, P3): The PR violates the ci.yml/ci-skip.yml path-filter invariant.
By C2 (P1, P2): This produces duplicate go-checks on invariants-only PRs.
By C3 (P4, P5, P6, P7, P9): The coverage script itself is correct.
By C4 (P7, P8): The CI job configuration is correct.
Therefore: One CRITICAL blocker (ci-skip.yml sync), one MEDIUM edge case (glob
  pattern tightening). The script logic and README documentation are sound.

CONFIDENCE: HIGH
- All premises verified by reading code and running regex tests
- No library/framework behavior assumed
- Alternative hypotheses checked and refuted
```

## Verification Checklist

- [x] Spot-check: P1 (`ci.yml:7-11`), P2 (`ci-skip.yml:6-9`), P6 (simulated script run)
- [x] Framework/library assumptions: None — pure Python stdlib + GitHub Actions path semantics
- [x] Alternative hypotheses ruled out: ci-skip.yml gap is real (no cross-workflow suppression); script has no false negatives on current state
- [ ] Unverified claims requiring running code: None — all verified statically
- [ ] Claims suitable for formal verification: None — this is CI/script infrastructure, not algorithmic logic


</details>

**Comment by hnipps** [comment]
**xylem — phase `test_critic` completed** (15.151572s)

_Usage:_ $0.0623, 3640

<details>
<summary>Phase output (click to expand)</summary>



## Test Critic Report

### Summary
- 0 test file(s) analyzed
- 0 finding(s) across 0 test function(s)
- Severity breakdown: 0 critical, 0 high, 0 medium, 0 low

**NO_TEST_FILES** — PR #665 contains no `*_test.go` files in its diff. The changes are a Python enforcement script, CI workflow updates, pre-commit config, and documentation. No Go test code to analyze for test theatre.

### Final Verdict
Not applicable — no test files in scope. The PR adds a coverage-checking tool rather than tests themselves.


</details>

**Comment by hnipps** [comment]
## Review Summary

**Verdict:** REQUEST_CHANGES

### Verification
One critical finding: `ci-skip.yml` `paths-ignore` not updated to match `ci.yml` `paths` additions (`docs/invariants/**`, `scripts/check_invariant_coverage.py`). This violates the documented CI path-filter invariant and will produce duplicate `go-checks` status contexts on invariants-only PRs.

The coverage script itself is correct — regex patterns, discovery logic, aspirational opt-out, and scoping are all sound. Verified against all 35 existing invariants (0 false positives, 0 false negatives).

### Test Quality
No test files changed — PR adds enforcement tooling, not Go tests.

### Inline Comments
2 inline comment(s) posted on specific lines.

**Comment by hnipps** [comment]
**xylem — phase `verify` completed** (7m44.724227s)

_Usage:_ $0.2253, 6359

<details>
<summary>Phase output (click to expand)</summary>

---

## Verification Report — PR #665

### Classification

CI/enforcement tooling (Python script + workflow config). Semi-formal `/reason` analysis.

---

### QUESTION
Are the code changes in PR #665 correct — does the invariant coverage script work as intended, and are the CI/pre-commit integrations sound?

### SCOPE
`scripts/check_invariant_coverage.py`, `.github/workflows/ci.yml`, `.github/workflows/ci-skip.yml`, `.pre-commit-config.yaml`, `docs/invariants/README.md`

---

### Premises

```
PREMISE P1 [STATIC]: ci.yml paths list has 5 entries after PR; ci-skip.yml paths-ignore has 3.
  Evidence: ci.yml:5-9 (PR commit 95210e5) — adds "docs/invariants/**", "scripts/check_invariant_coverage.py"
  Evidence: ci-skip.yml:6-9 (unchanged) — only "cli/**", ".golangci.yml", ".xylem/workflows/**"

PREMISE P2 [STATIC]: CLAUDE.md mandates these lists be identical at all times.
  Evidence: CLAUDE.md § "CI path-filter invariant"

PREMISE P3 [SEMANTIC]: GitHub Actions paths-ignore fires the workflow when ANY file does NOT match ALL patterns.
  Evidence: GitHub Actions docs — if a changed file is not in paths-ignore, the workflow runs.

PREMISE P4 [STATIC]: ci-skip.yml defines only one job: go-checks. ci.yml now defines two: go-checks + invariant-coverage.
  Evidence: ci-skip.yml:11-15 — single go-checks job
  Evidence: ci.yml:49-55 (PR commit 95210e5) — new invariant-coverage job

PREMISE P5 [BEHAVIORAL]: Script runs correctly on current codebase: 35/35 required invariants covered, exit 0.
  Evidence: Local execution — "invariant coverage OK: 35/35 required invariants covered across 3 module(s) (0 aspirational)."

PREMISE P6 [STATIC]: HEADER_RE matches all current invariant headers (I1-I14, I1a, I5a, I5b, S1-S8).
  Evidence: Grep of docs/invariants/*.md — 35 headers, all matched by ^\*\*([A-Z]+\d+[a-z]?)\.\s

PREMISE P7 [STATIC]: TEST_COMMENT_RE matches all current test comments.
  Evidence: Grep of cli/internal/**/*invariants_prop_test.go — 35 comments, all matched by ^//\s*Invariant\s+([A-Z]+\d+[a-z]?):\s

PREMISE P8 [STATIC]: No invariant doc currently uses <!-- aspirational --> HTML comment.
  Evidence: Grep for "aspirational\s*-->" in docs/invariants/ — no matches.

PREMISE P9 [STATIC]: REPO_ROOT = Path(__file__).resolve().parent.parent resolves correctly for both pre-commit (CWD=repo root) and CI (actions/checkout CWD).
  Evidence: scripts/check_invariant_coverage.py:43
```

---

### Claims

```
CLAIM C1 [STATIC]: The ci.yml / ci-skip.yml path-filter invariant is violated.
  Trace: ci.yml paths has 5 entries, ci-skip.yml paths-ignore has 3
  Depends on: P1, P2

CLAIM C2 [SEMANTIC]: CORRECTION — the initial analysis claim that "neither workflow fires" is wrong.
  For invariant-only changes, BOTH fire.
  Trace: PR touches only docs/invariants/foo.md →
    ci.yml: "docs/invariants/**" matches → TRIGGERS
    ci-skip.yml: file NOT in paths-ignore → TRIGGERS
  Depends on: P1, P3

CLAIM C3 [SEMANTIC]: The practical consequence is wasted CI (Go build+test runs unnecessarily on non-Go changes), not blocked merges.
  Depends on: C2

CLAIM C4 [STATIC]: ci-skip.yml lacks a no-op invariant-coverage job.
  If invariant-coverage is later made a required status check, non-Go-non-invariant PRs will be permanently blocked (ci.yml doesn't fire, ci-skip.yml fires but provides only go-checks, not invariant-coverage).
  Depends on: P4

CLAIM C5 [BEHAVIORAL]: The Python script is correct for all existing invariant specs and test files.
  Trace: discover_modules() finds queue/runner/scanner →
    load_doc() extracts 35 IDs → load_tests() finds 35 matching comments →
    required - aspirational = 35 - 0 = 35 → covered ∩ required = 35 → pass
  Depends on: P5, P6, P7, P8

CLAIM C6 [STATIC]: Pre-commit hook file pattern is correct and complete.
  Evidence: .pre-commit-config.yaml:68 — ^(docs/invariants/.+\.md|cli/internal/.+_invariants_prop_test\.go|scripts/check_invariant_coverage\.py)$
  Triggers on doc changes, test file changes, and script changes. Correct.
  Depends on: P6, P7
```

---

### Alternative Hypothesis Check

```
ALTERNATIVE HYPOTHESIS CHECK:
"The ci-skip.yml path mismatch will block PRs"
- Searched for: scenarios where NEITHER workflow fires (which would leave go-checks pending)
- Found: For any file NOT in ci.yml paths AND NOT in ci-skip.yml paths-ignore: ci.yml doesn't fire, ci-skip.yml DOES fire (provides no-op go-checks). For invariant-only files: ci.yml DOES fire. No scenario leaves go-checks missing.
- Conclusion: REFUTED — no merge-blocking scenario exists for go-checks.

"The script has a false-negative for aspirational invariants"
- Searched for: invariants described as "aspirational" in prose but lacking <!-- aspirational --> HTML comment
- Found: I5b (queue.md:97), I7 (runner.md:169), I13 (runner.md:328) — all have "aspirational" in prose but no HTML comment
- All three HAVE test comments (queue_invariants_prop_test.go:430, runner_invariants_prop_test.go:325, runner_invariants_prop_test.go:629)
- Conclusion: REFUTED — the script treats them as required and they are covered. No false negative.
```

---

### Formal Conclusion

By C1 (P1, P2): The stated ci.yml/ci-skip.yml path-filter invariant is violated.

By C2, C3 (P1, P3): The consequence is wasted CI minutes (both workflows fire for invariant-only changes), not blocked merges. The initial analysis's "permanently pending" claim is incorrect.

By C4 (P4): A latent blocking risk exists if `invariant-coverage` is later made a required status check without adding a matching no-op to ci-skip.yml.

By C5, C6 (P5-P8): The Python script and pre-commit integration are correct and functional.

**CONFIDENCE: HIGH** — All premises verified by reading code, running the script, and tracing GitHub Actions semantics.

---

## Findings

### Finding 1: ci-skip.yml paths-ignore not synchronized

| Attribute | Value |
|-----------|-------|
| **File** | `.github/workflows/ci-skip.yml:6-9` |
| **Severity** | HIGH |
| **Category** | Invariant violation |

**Description**: The PR adds `docs/invariants/**` and `scripts/check_invariant_coverage.py` to `ci.yml`'s `paths:` list but does not add matching entries to `ci-skip.yml`'s `paths-ignore:` list. This violates the stated invariant in CLAUDE.md. The initial analysis's claim that this would cause "neither workflow to fire" is incorrect — in practice, both workflows fire on invariant-only PRs, causing unnecessary Go build/test execution (~2-5 min wasted CI per PR).

**Evidence**: `ci.yml:5-9` (5 entries) vs `ci-skip.yml:6-9` (3 entries) in PR commit `95210e5`.

**Fix**: Add `"docs/invariants/**"` and `"scripts/check_invariant_coverage.py"` to `ci-skip.yml`'s `paths-ignore:` list.

---

### Finding 2: ci-skip.yml missing `invariant-coverage` no-op job

| Attribute | Value |
|-----------|-------|
| **File** | `.github/workflows/ci-skip.yml` |
| **Severity** | MEDIUM |
| **Category** | Edge case |

**Description**: `ci.yml` now defines two jobs (`go-checks` and `invariant-coverage`) but `ci-skip.yml` only provides a no-op for `go-checks`. If `invariant-coverage` is ever made a required status check, PRs that don't trigger `ci.yml` (e.g., changes to only `README.md` at root) will be permanently blocked — `ci-skip.yml` fires but doesn't provide the `invariant-coverage` check.

**Evidence**: `ci.yml:49-55` defines `invariant-coverage` job; `ci-skip.yml:11-15` only defines `go-checks`.

**Fix**: Add a no-op `invariant-coverage` job to `ci-skip.yml`, or document that `invariant-coverage` must not be made a required check.

---

### Finding 3 (informational): `sys.exit(2)` vs `return 2` inconsistency

| Attribute | Value |
|-----------|-------|
| **File** | `scripts/check_invariant_coverage.py:77` vs `:121` |
| **Severity** | LOW (informational) |
| **Category** | Correctness (cosmetic) |

**Description**: `discover_modules()` calls `sys.exit(2)` directly when `DOCS_DIR` is missing, while `main()` returns `2` for the "no modules found" case. Both exit with code 2 but the mechanisms differ — `sys.exit` bypasses `main()`'s return and any future cleanup code.

---

### Verification Checklist

- [x] Spot-check P1: ci.yml and ci-skip.yml path lists (`ci.yml:5-9`, `ci-skip.yml:6-9`)
- [x] Spot-check P5: Script execution produces 35/35 pass
- [x] Framework assumptions: GitHub Actions `paths-ignore` semantics (P3)
- [x] Alternative hypotheses ruled out: "blocks merge" (REFUTED), "aspirational false-negative" (REFUTED)
- [x] No `[BEHAVIORAL]` claims remain unverified — script was executed
- [ ] No `[FORMAL]` items warranting Dafny verification (text-processing script, not safety-critical)


</details>

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -3,10 +3,20 @@ name: CI
 on:
   push:
     branches: [main]
-    paths: ["cli/**", ".golangci.yml", ".xylem/workflows/**"]
+    paths:
+      - "cli/**"
+      - ".golangci.yml"
+      - ".xylem/workflows/**"
+      - "docs/invariants/**"
+      - "scripts/check_invariant_coverage.py"
```
Comment (hnipps) [inline .github/workflows/ci.yml:11]: **BLOCKER**: `ci-skip.yml` `paths-ignore` must be updated to include `docs/invariants/**` and `scripts/check_invariant_coverage.py` — the two new entries added here.

Per the documented CI path-filter invariant (CLAUDE.md § "CI path-filter invariant"): the `paths:` list in `ci.yml` and the `paths-ignore:` list in `ci-skip.yml` must be identical at all times.

Without this fix, a PR touching only `docs/invariants/**` will fire **both** `ci.yml` (path matches) and `ci-skip.yml` (path not excluded), producing duplicate `go-checks` status contexts.

File: scripts/check_invariant_coverage.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+
+def discover_modules() -> list[ModuleCoverage]:
+    """Locate (module, doc, test_paths) triples for every invariant doc."""
+    if not DOCS_DIR.is_dir():
+        print(f"ERROR: {DOCS_DIR} is missing", file=sys.stderr)
+        sys.exit(2)
+
+    modules: list[ModuleCoverage] = []
+    for doc in sorted(DOCS_DIR.glob("*.md")):
```
Comment (hnipps) [inline scripts/check_invariant_coverage.py:80]: Minor: this glob `*invariants_prop_test.go` would also match a hypothetical `invariants_prop_test.go` (no module prefix), while the pre-commit `files` regex requires `_invariants_prop_test.go` (with leading underscore). Consider tightening to `*_invariants_prop_test.go` for consistency.

No existing files trigger this — all 3 current files have module prefixes — so this is non-blocking.

**Review by hnipps — COMMENTED** [review]
See inline comments.

**Comment by hnipps** [comment]
**xylem — phase `review` completed** (2m28.826842s)

_Usage:_ $0.0920, 5742

<details>
<summary>Phase output (click to expand)</summary>

Review posted on PR #665:

- **Anchor comment** with summary and REQUEST_CHANGES verdict
- **2 inline comments**: (1) BLOCKER on `ci.yml:11` — `ci-skip.yml` paths-ignore must be synced with the 2 new path entries, (2) minor suggestion on `check_invariant_coverage.py:80` — tighten glob from `*invariants_prop_test.go` to `*_invariants_prop_test.go` for consistency with the pre-commit regex


</details>

**Comment by hnipps** [comment]
**xylem — all phases completed**

| Phase | Duration | Cost | Tokens | Status |
|-------|----------|------|--------|--------|
| analyze | 4m20.499866s | $0.0941 | 4205 | completed |
| verify | 5m24.543361s | $0.2105 | 6146 | completed |
| test_critic | 15.151572s | $0.0623 | 3640 | completed |
| review | 2m28.826842s | $0.0920 | 5742 | completed |

Total: 12m29.021641s — $0.4588, 19733 tokens (estimated)

**Comment by hnipps** [comment]
**xylem — phase `test_critic` completed** (16.416037s)

_Usage:_ $0.0663, 3708

<details>
<summary>Phase output (click to expand)</summary>



## Test Critic Report

### Summary

NO_TEST_FILES

No `*_test.go` files were changed in PR #665. The PR adds a Python enforcement script (`check_invariant_coverage.py`), CI wiring, and documentation — no Go test files are in scope for test-theatre analysis.

### Notable Non-Test Finding

The CI path-filter invariant violation identified in the analysis is the actionable issue on this PR: `docs/invariants/**` and `scripts/check_invariant_coverage.py` are added to `ci.yml` `paths:` but missing from `ci-skip.yml` `paths-ignore:`, which will cause the required `go-checks` status to hang on PRs touching only those paths.

### Final Verdict

No test theatre to report — there are no test files to critique.


</details>

**Comment by hnipps** [comment]
**xylem — phase `review` completed** (1m31.12634s)

_Usage:_ $0.1397, 6589

<details>
<summary>Phase output (click to expand)</summary>

Both GitHub tokens are invalid. I cannot post review comments — `gh` auth is completely broken this session (both `GH_TOKEN` and the stored `hnipps` credential are expired/invalid).

Here are the review comments ready to post manually:

---

**Anchor comment** (post via `gh pr comment 665`):

```
<!-- xylem-review -->
## Review Summary

**Verdict:** REQUEST_CHANGES

### Verification
- **HIGH**: `ci-skip.yml` `paths-ignore` not synchronized with `ci.yml` `paths` — violates the stated CI path-filter invariant (CLAUDE.md). Two new entries (`docs/invariants/**`, `scripts/check_invariant_coverage.py`) added to `ci.yml` but missing from `ci-skip.yml`. Practical impact: both workflows fire on invariant-only PRs, wasting CI minutes on unnecessary Go builds.
- **MEDIUM**: `ci-skip.yml` lacks a no-op `invariant-coverage` job. If this check is ever made required, non-Go/non-invariant PRs will be permanently blocked.
- Script logic verified correct: 35/35 invariants covered, regex patterns match all existing headers and test comments.

### Test Quality
No test files changed — enforcement is via CI + pre-commit.

### Inline Comments
2 inline comment(s) posted on specific lines.
```

**Inline comment 1** — `.github/workflows/ci-skip.yml`, line 9:

```
ci.yml now has 5 entries in `paths:` but this file still has 3 in `paths-ignore:`. Missing:
- `docs/invariants/**`
- `scripts/check_invariant_coverage.py`

Per CLAUDE.md: "the `paths:` list in `ci.yml` and the `paths-ignore:` list in `ci-skip.yml` must be identical at all times."

**What breaks**: invariant-only PRs trigger *both* workflows — `ci.yml` fires (file matches paths) and `ci-skip.yml` also fires (file not in paths-ignore). The Go build/test suite runs unnecessarily on non-Go changes.
```

**Inline comment 2** — `.github/workflows/ci-skip.yml`, line 15:

```
`ci.yml` now defines two jobs (`go-checks` + `invariant-coverage`) but this file only provides a no-op for `go-checks`.

**What breaks**: if `invariant-coverage` is ever made a required status check, PRs that don't touch Go or invariant files (e.g., root README changes) will be permanently blocked — `ci-skip.yml` fires but only satisfies `go-checks`, not `invariant-coverage`.

Fix: add a matching no-op `invariant-coverage` job here, or document that this check must not be made required.
```

To post these when auth is restored:
```bash
gh pr comment 665 --body '<anchor comment above>'
gh api repos/nicholls-inc/xylem/pulls/665/reviews --method POST \
  -f event="COMMENT" \
  --input <(echo '{"comments":[{"path":".github/workflows/ci-skip.yml","line":9,"side":"RIGHT","body":"..."},{"path":".github/workflows/ci-skip.yml","line":15,"side":"RIGHT","body":"..."}]}')
```


</details>

**Comment by hnipps** [comment]
**xylem — all phases completed**

| Phase | Duration | Cost | Tokens | Status |
|-------|----------|------|--------|--------|
| analyze | 2m47.212349s | $0.0954 | 4223 | completed |
| verify | 7m44.724227s | $0.2253 | 6359 | completed |
| test_critic | 16.416037s | $0.0663 | 3708 | completed |
| review | 1m31.12634s | $0.1397 | 6589 | completed |

Total: 12m19.478953s — $0.5268, 20879 tokens (estimated)
